internal/provider/kubernetes/predicates.go

```diff
@@ -226,9 +229,15 @@ func (r *gatewayAPIReconciler) validateServiceForReconcile(obj client.Object) bo
 		res, _ := r.resources.GatewayAPIResources.Load(string(r.classController))
 		if res != nil {
 			if (*res)[gclass] != nil && len((*res)[gclass].Gateways) > 0 {
-				for _, gw := range (*res)[gclass].Gateways {
-					gw := gw
-					r.updateStatusForGateway(ctx, gw)
+				for _, gtw := range (*res)[gclass].Gateways {
+					curGtw := new(gwapiv1.Gateway)
+					key := types.NamespacedName{Namespace: gtw.Namespace, Name: gtw.Name}
+					if err := r.client.Get(ctx, key, curGtw); err != nil {
+						r.log.Info("gateway not found", "gatewayclass", gclass, "name", key.String())
+						return false
+					}
+
+					r.updateStatusForGateway(ctx, curGtw)
 				}
 			}
 		}
```

Comment:
I was more thinking about something like this, wdyt?
```bash
	gclass, ok := labels[gatewayapi.OwningGatewayClassLabel]
	if ok && r.mergeGateways[gclass] {
		gatewayList := &gwapiv1.GatewayList{}
		if err := r.client.List(ctx, gatewayList, &client.ListOptions{
			FieldSelector: fields.OneTermEqualSelector(classGatewayIndex, gclass),
		}); err != nil {
			r.log.Info("no associated Gateways found for GatewayClass", "name", gclass)
		}
		if len(gatewayList.Items) > 0 {
			for _, gw := range gatewayList.Items {
				gw := gw
				r.updateStatusForGateway(ctx, &gw)
			}

		}
		return false
	}
```

Reply:
we could move it also to separate func like r.findOwningGateways

Reply:
@cnvergence one question: for example, if we got A,B,C three gateways, and the listener conflict only happened between B and C. what we expected to see:
- (1) A, B should works but C should not;
- (2) neither of A, B, C should work;

Reply:
we should set the option A, meaning that the newest gateway with conflict will get rejected

Reply:
this is expected behaviour https://github.com/envoyproxy/gateway/blob/main/internal/gatewayapi/testdata/merge-invalid-multiple-gateways.out.yaml

Reply:
updated, thanks for answering